File: DSAL/A-CPP/A4.cpp
// 4. To create ADT that implement the "set" concept.
//   1. Add (New Element) Place a value into the set
//   2. Remove (Element) Remove the value
//   3. Contains (Element) Return true if element is in collection
//   4. Size () Return number of values in collection
//   5. Iterator () Return an iterator used to loop over collection
//   6. Intersection of two sets
//   7. Union of two sets
//   8. Difference between two sets
//   9. Subset

#include <iostream>
#include <string>
using namespace std;

class SetIterator {
 private:
  int *arr;
  int count;
  int idx;

 public:
  SetIterator(int *data, int n) {
    arr = data;
    count = n;
    idx = 0;
  }

  bool hasNext() const { return idx < count; }

  int next() {
    if (!hasNext()) return 0;
    return arr[idx++];
  }
};

class SetADT {
 private:
  int *data;
  int count;
  int capacity;

  void ensureCapacity() {
    if (count < capacity) return;

    int newCap = capacity * 2;
    if (newCap < 4) newCap = 4;

    int *newData = new int[newCap];
    for (int i = 0; i < count; ++i) {
      newData[i] = data[i];
    }
    delete[] data;
    data = newData;
    capacity = newCap;
  }

  int binarySearch(int key) const {
    int low = 0;
    int high = count - 1;
    while (low <= high) {
      int mid = low + (high - low) / 2;
      if (data[mid] == key) return mid;
      if (data[mid] < key) low = mid + 1;
      else high = mid - 1;
    }
    return -1;
  }

  int findInsertPosition(int key) const {
    int low = 0;
    int high = count;
    while (low < high) {
      int mid = low + (high - low) / 2;
      if (data[mid] < key) low = mid + 1;
      else high = mid;
    }
    return low;
  }

 public:
  SetADT(int cap = 8) {
    if (cap < 1) cap = 8;
    capacity = cap;
    count = 0;
    data = new int[capacity];
  }

  SetADT(const SetADT &other) {
    capacity = other.capacity;
    count = other.count;
    data = new int[capacity];
    for (int i = 0; i < count; ++i) {
      data[i] = other.data[i];
    }
  }

  SetADT &operator=(const SetADT &other) {
    if (this == &other) return *this;
    delete[] data;
    capacity = other.capacity;
    count = other.count;
    data = new int[capacity];
    for (int i = 0; i < count; ++i) {
      data[i] = other.data[i];
    }
    return *this;
  }

  ~SetADT() { delete[] data; }

  bool add(int element) {
    if (contains(element)) return false;
    ensureCapacity();
    int pos = findInsertPosition(element);
    for (int i = count; i > pos; --i) {
      data[i] = data[i - 1];
    }
    data[pos] = element;
    ++count;
    return true;
  }

  bool remove(int element) {
    int idx = binarySearch(element);
    if (idx == -1) return false;
    for (int i = idx; i < count - 1; ++i) {
      data[i] = data[i + 1];
    }
    --count;
    return true;
  }

  bool contains(int element) const { return binarySearch(element) != -1; }

  int size() const { return count; }

  SetIterator iterator() { return SetIterator(data, count); }

  void display() const {
    cout << "{ ";
    for (int i = 0; i < count; ++i) {
      cout << data[i] << " ";
    }
    cout << "}\n";
  }

  SetADT unionSet(const SetADT &other) const {
    SetADT result(count + other.count + 2);
    int i = 0, j = 0;
    while (i < count && j < other.count) {
      if (data[i] < other.data[j]) result.data[result.count++] = data[i++];
      else if (data[i] > other.data[j]) result.data[result.count++] = other.data[j++];
      else {
        result.data[result.count++] = data[i];
        ++i;
        ++j;
      }
    }
    while (i < count) result.data[result.count++] = data[i++];
    while (j < other.count) result.data[result.count++] = other.data[j++];
    return result;
  }

  SetADT intersection(const SetADT &other) const {
    SetADT result((count < other.count ? count : other.count) + 2);
    int i = 0, j = 0;
    while (i < count && j < other.count) {
      if (data[i] < other.data[j]) ++i;
      else if (data[i] > other.data[j]) ++j;
      else {
        result.data[result.count++] = data[i];
        ++i;
        ++j;
      }
    }
    return result;
  }

  SetADT difference(const SetADT &other) const {
    SetADT result(count + 2);
    int i = 0, j = 0;
    while (i < count && j < other.count) {
      if (data[i] < other.data[j]) result.data[result.count++] = data[i++];
      else if (data[i] > other.data[j]) ++j;
      else {
        ++i;
        ++j;
      }
    }
    while (i < count) result.data[result.count++] = data[i++];
    return result;
  }

  bool isSubsetOf(const SetADT &other) const {
    int i = 0, j = 0;
    while (i < count && j < other.count) {
      if (data[i] < other.data[j]) return false;
      if (data[i] > other.data[j]) ++j;
      else {
        ++i;
        ++j;
      }
    }
    return i == count;
  }
};

void setMenu(SetADT &s, const string &name) {
  while (true) {
    cout << "\n"
         << name << " Menu\n";
    cout << "1. Add element\n";
    cout << "2. Remove element\n";
    cout << "3. Contains element\n";
    cout << "4. Size\n";
    cout << "5. Display\n";
    cout << "6. Iterate\n";
    cout << "7. Back\n";
    cout << "Enter choice: ";

    int choice;
    cin >> choice;

    if (choice == 1) {
      int x;
      cout << "Enter element: ";
      cin >> x;
      cout << (s.add(x) ? "Added\n" : "Duplicate element\n");
    } else if (choice == 2) {
      int x;
      cout << "Enter element: ";
      cin >> x;
      cout << (s.remove(x) ? "Removed\n" : "Not found\n");
    } else if (choice == 3) {
      int x;
      cout << "Enter element: ";
      cin >> x;
      cout << (s.contains(x) ? "Present\n" : "Absent\n");
    } else if (choice == 4) {
      cout << "Size: " << s.size() << "\n";
    } else if (choice == 5) {
      s.display();
    } else if (choice == 6) {
      SetIterator it = s.iterator();
      cout << "Iterator output: ";
      while (it.hasNext()) cout << it.next() << " ";
      cout << "\n";
    } else if (choice == 7) {
      break;
    } else {
      cout << "Invalid choice\n";
    }
  }
}

int main() {
  SetADT setA, setB;

  while (true) {
    cout << "\nMain Menu\n";
    cout << "1. Operate on Set A\n";
    cout << "2. Operate on Set B\n";
    cout << "3. Union (A U B)\n";
    cout << "4. Intersection (A n B)\n";
    cout << "5. Difference (A - B)\n";
    cout << "6. Difference (B - A)\n";
    cout << "7. Is A subset of B?\n";
    cout << "8. Display both sets\n";
    cout << "9. Exit\n";
    cout << "Enter choice: ";

    int choice;
    cin >> choice;

    if (choice == 1) {
      setMenu(setA, "Set A");
    } else if (choice == 2) {
      setMenu(setB, "Set B");
    } else if (choice == 3) {
      SetADT r = setA.unionSet(setB);
      cout << "Union: ";
      r.display();
    } else if (choice == 4) {
      SetADT r = setA.intersection(setB);
      cout << "Intersection: ";
      r.display();
    } else if (choice == 5) {
      SetADT r = setA.difference(setB);
      cout << "Difference (A-B): ";
      r.display();
    } else if (choice == 6) {
      SetADT r = setB.difference(setA);
      cout << "Difference (B-A): ";
      r.display();
    } else if (choice == 7) {
      if (setA.isSubsetOf(setB)) cout << "Set A is a subset of Set B\n";
      else cout << "Set A is NOT a subset of Set B\n";
    } else if (choice == 8) {
      cout << "Set A: ";
      setA.display();
      cout << "Set B: ";
      setB.display();
    } else if (choice == 9) {
      cout << "Exiting...\n";
      break;
    } else {
      cout << "Invalid choice\n";
    }
  }

  return 0;
}

File: DSAL/A/A4.py
# 4. To create ADT that implement the "set" concept.
#   1. Add (New Element) Place a value into the set
#   2. Remove (Element) Remove the value
#   3. Contains (Element) Return true if element is in collection
#   4. Size () Return number of values in collection
#   5. Iterator () Return an iterator used to loop over collection
#   6. Intersection of two sets
#   7. Union of two sets
#   8. Difference between two sets
#   9. Subset

class SetIterator:
    def __init__(self, data, count):
        self.data = data
        self.count = count
        self.current = 0

    def has_next(self):
        return self.current < self.count

    def next(self):
        if self.current < self.count:
            val = self.data[self.current]
            self.current += 1
            return val
        return None

class SetADT:
    def __init__(self, capacity=100):
        self.capacity = capacity
        self.data = [None] * capacity
        self.count = 0

    def _binary_search(self, key):
        low = 0
        high = self.count - 1
        while low <= high:
            mid = (low + high) // 2
            if self.data[mid] == key:
                return mid
            elif self.data[mid] < key:
                low = mid + 1
            else:
                high = mid - 1
        return -1

    def _find_insert_position(self, key):
        low = 0
        high = self.count - 1
        while low <= high:
            mid = (low + high) // 2
            if self.data[mid] < key:
                low = mid + 1
            else:
                high = mid - 1
        return low

    def add(self, element):
        if self.count >= self.capacity:
            print("Set is full")
            return False

        if self.contains(element):
            return False

        pos = self._find_insert_position(element)

        i = self.count
        while i > pos:
            self.data[i] = self.data[i - 1]
            i -= 1

        self.data[pos] = element
        self.count += 1
        return True

    def remove(self, element):
        index = self._binary_search(element)
        if index == -1:
            return False

        i = index
        while i < self.count - 1:
            self.data[i] = self.data[i + 1]
            i += 1

        self.data[self.count - 1] = None
        self.count -= 1
        return True

    def contains(self, element):
        return self._binary_search(element) != -1

    def size(self):
        return self.count

    def iterator(self):
        return SetIterator(self.data, self.count)

    def display(self):
        print("{", end=" ")
        for i in range(self.count):
            print(self.data[i], end=" ")
        print("}")

    def union(self, other_set):
        result = SetADT(self.capacity + other_set.capacity)
        i = 0
        j = 0

        while i < self.count and j < other_set.count:
            if self.data[i] < other_set.data[j]:
                result.add(self.data[i])
                i += 1
            elif self.data[i] > other_set.data[j]:
                result.add(other_set.data[j])
                j += 1
            else:
                result.add(self.data[i])
                i += 1
                j += 1

        while i < self.count:
            result.add(self.data[i])
            i += 1

        while j < other_set.count:
            result.add(other_set.data[j])
            j += 1

        return result

    def intersection(self, other_set):
        result = SetADT(min(self.capacity, other_set.capacity))
        i = 0
        j = 0

        while i < self.count and j < other_set.count:
            if self.data[i] < other_set.data[j]:
                i += 1
            elif self.data[i] > other_set.data[j]:
                j += 1
            else:
                result.add(self.data[i])
                i += 1
                j += 1

        return result

    def difference(self, other_set):
        result = SetADT(self.capacity)
        i = 0
        j = 0

        while i < self.count and j < other_set.count:
            if self.data[i] < other_set.data[j]:
                result.add(self.data[i])
                i += 1
            elif self.data[i] > other_set.data[j]:
                j += 1
            else:
                i += 1
                j += 1

        while i < self.count:
            result.add(self.data[i])
            i += 1

        return result

    def subset(self, other_set):
        i = 0
        j = 0

        while i < self.count and j < other_set.count:
            if self.data[i] < other_set.data[j]:
                return False
            elif self.data[i] > other_set.data[j]:
                j += 1
            else:
                i += 1
                j += 1

        return i == self.count

def main():
    setA = SetADT()
    setB = SetADT()

    while True:
        print("\n1. Add to Set A")
        print("2. Add to Set B")
        print("3. Remove from Set A")
        print("4. Remove from Set B")
        print("5. Contains in Set A")
        print("6. Size of Sets")
        print("7. Iterator (Display Set A)")
        print("8. Union (A U B)")
        print("9. Intersection (A n B)")
        print("10. Difference (A - B)")
        print("11. Subset (A subset of B?)")
        print("12. Display Both Sets")
        print("13. Exit")

        choice = input("Enter choice: ")

        if choice == '1':
            val = int(input("Enter element: "))
            if setA.add(val):
                print("Added")
            else:
                print("Duplicate or Full")
        elif choice == '2':
            val = int(input("Enter element: "))
            if setB.add(val):
                print("Added")
            else:
                print("Duplicate or Full")
        elif choice == '3':
            val = int(input("Enter element: "))
            if setA.remove(val):
                print("Removed")
            else:
                print("Not Found")
        elif choice == '4':
            val = int(input("Enter element: "))
            if setB.remove(val):
                print("Removed")
            else:
                print("Not Found")
        elif choice == '5':
            val = int(input("Enter element: "))
            if setA.contains(val):
                print("Found")
            else:
                print("Not Found")
        elif choice == '6':
            print("Size A:", setA.size())
            print("Size B:", setB.size())
        elif choice == '7':
            it = setA.iterator()
            print("Iterating Set A:")
            while it.has_next():
                print(it.next(), end=" ")
            print()
        elif choice == '8':
            res = setA.union(setB)
            print("Union:", end=" ")
            res.display()
        elif choice == '9':
            res = setA.intersection(setB)
            print("Intersection:", end=" ")
            res.display()
        elif choice == '10':
            res = setA.difference(setB)
            print("Difference (A-B):", end=" ")
            res.display()
        elif choice == '11':
            if setA.subset(setB):
                print("Set A is a subset of Set B")
            else:
                print("Set A is NOT a subset of Set B")
        elif choice == '12':
            print("Set A:", end=" ")
            setA.display()
            print("Set B:", end=" ")
            setB.display()
        elif choice == '13':
            break
        else:
            print("Invalid choice")

main()